### ISSUE TITLE
Push for Bert, MaskRCNN and Resnet-18 E2E support using Lazy Tensor Core

### ISSUE THREAD

**Issue by ramiro050** [llvm/torch-mlir#365]
The purpose of this bug is to help keep track of tasks needed to have E2E support for `Bert`, `MaskRCNN`, and `Resnet-18` using the Lazy Tensor Core (LTC) and lowering through `torch-mlir`.

At the moment, the main part that is missing is op support in LTC. Below is a table of the LTC ops needed. The list of ops was determined by running two scripts: this [MaskRCNN script](https://github.com/ramiro050/lazy-tensor-samples/blob/main/lazytensor_maskrcnn_example.py) which generates this [output](https://github.com/ramiro050/lazy-tensor-samples/blob/main/lazytensor_maskrcnn_example_output.txt), and this [Resnet script](https://github.com/ramiro050/lazy-tensor-samples/blob/main/lazytensor_resnet18_example.py) which generates this [output](https://github.com/ramiro050/lazy-tensor-samples/blob/main/lazytensor_resnet18_example_output.txt). The ops missing are the ones with the `aten::` prefix in the output of the scripts. For more information of how to setup and run the examples, see [here](https://github.com/ramiro050/lazy-tensor-samples/blob/main/README.md).

### LTC Ops Needed

Status column symbols:
`-` unimplemented
`+` started
`x` finished

| Ops                             | Status | Owner     | Model            | Notes |
|---------------------------------|--------|-----------|------------------|-------|
| `aten::_index_put_impl_`        | -      |           | MaskRCNN         |       |
| `aten::arange.start_out`        | +      |     silvasean      | MaskRCNN         |       |
| `aten::exp.out`                 | x      | ramiro050          | MaskRCNN         | https://github.com/pytorch/pytorch/pull/67213      |
| `aten::floor.out`               | x      | ramiro050          | MaskRCNN         | https://github.com/pytorch/pytorch/pull/66770      |
| `aten::index.Tensor`            | +      | ramiro050          | MaskRCNN         |       |
| `aten::log2.out`                | x      | ramiro050          | MaskRCNN         | https://github.com/pytorch/pytorch/pull/66771      |
| `aten::max_pool2d_with_indices` | +      |   vivekkhandelwal1        | MaskRCNN, Resnet |       |
| `aten::upsample_nearest2d.out`  | -      |           | MaskRCNN         |       |
| `aten::mean.out`                | x      | alanwaketan | Resnet           | https://github.com/pytorch/pytorch/pull/67174      |
| `aten::sort`                    | x      |  silvasean  | Resnet           |   https://github.com/pytorch/pytorch/pull/67053    |

## torch-mlir ops Needed

Below is a list of ops needed on the `torch-mlir` side. This list was compiled by going over the ops detected by LTC when running this [MaskRCNN script](https://github.com/ramiro050/lazy-tensor-samples/blob/main/lazytensor_maskrcnn_example.py) (output with list of ops detected can be found [here](https://github.com/ramiro050/lazy-tensor-samples/blob/main/lazytensor_maskrcnn_example_output.txt)), this [Bert script](https://github.com/ramiro050/lazy-tensor-samples/blob/main/lazytensor_bert_example.py) ([output](https://github.com/ramiro050/lazy-tensor-samples/blob/main/lazytensor_bert_example_output.txt)), and the Resnet-18 model in the PyTorch benchmarks ([instructions](https://github.com/ramiro050/lazy-tensor-samples#resnet-18-inference-and-training) for setting it up with LTC), and checking which had lowerings in `torch-mlir` and which did not. 

**Note:** `Bert` and `Resnet18` are currently the only training models. The ops needed for `MaskRCNN` training will be added soon.

Status column symbols:
`-` unimplemented
`+` started
`x` finished

The full op lists including finished ones are moved to https://github.com/llvm/torch-mlir/issues/365#issuecomment-1022664476. This new table only contains ops to be done so that we can be more focused.
| Op                                        | Status | Owner    | Model                                                 | Notes |
|-------------------------------------------|--------|----------|-------------------------------------------------------|-------|
| `aten::bernoulli_`                        | +      |    pashu123      | Bert Training                                         |   rng op |
| `aten::embedding_dense_backward`          | +      | vivekkhandelwal1         | Bert Training                                         |   histogram    |
| `aten::native_layer_norm_backward`        | +    |  gprateek93        | Bert Training                                         |   [PR546](https://github.com/llvm/torch-mlir/pull/546), [PR570](https://github.com/llvm/torch-mlir/pull/570)   |
| `aten::nll_loss_backward`                 | +      |   pashu123       | Bert Training, Resnet-18 Training                     | [PR463](https://github.com/llvm/torch-mlir/pull/463)      |
| `aten::_copy_from`                        | +      | pashu123   | Resnet-18 Training, MaskRCNN Inference | torchscript baseline won't run (to be investigated)|
| `aten::convolution_backward_overrideable` | +      | gpetters94         | Resnet-18 Training                                    |       |
| `aten::max_pool2d_with_indices_backward`  | +      |    vivekkhandelwal1      | Resnet-18 Training                                    |       |
| `aten::native_batch_norm_backward`        | +      |   Shukla-Gaurav       | Resnet-18 Training                                    |       |
| `aten::native_batch_norm`                 | +      |  Shukla-Gaurav        | Resnet-18 Training                                    | [PR563](https://github.com/llvm/torch-mlir/pull/563)      |
| `aten::random_.to`                        | +      |     gprateek93     | Resnet-18 Training                                    |  rng op     |
| `aten::_copy_from_and_resize`             | +      |  gpetters94       | Resnet-18 Training, MaskRCNN Inference                |       |
| `aten::convolution_overrideable`          | +      |  gpetters94       | Resnet-18 Training, MaskRCNN Inference                |       |
| `aten::convolution`          | +    |  gpetters94        | Resnet-18 Training(through AOTAutograd)                |       |
| `aten::convolution_backward`          | +      |  gpetters94       | Resnet-18 Training(through AOTAutograd)                |       |
| `aten::max_pool2d_with_indices`           | +     | vivekkhandelwal1         | Resnet-18 Training, MaskRCNN Inference                | https://github.com/llvm/torch-mlir/pull/518      |
| `aten::_index_put_impl_`                  | -      |          | MaskRCNN Inference                                    |   histogram    |
| `aten::stack`                             | -      |  pashu123        | MaskRCNN Inference                                    |       |
| `aten::topk`                              | -      |    gprateek93      | MaskRCNN Inference                                    |       |
| `aten::upsample_nearest2d`                | -      |     gprateek93     | MaskRCNN Inference                                    |       |
| `torchvision::nms`                        | -      |          | MaskRCNN Inference                                    |       |
| `torchvision::roi_align`                  | -      |          | MaskRCNN Inference                                    |       |

**Comment by silvasean** [comment]
Awesome. Just as an FYI, I'm working on plugging torch-mlir into LTC.

**Comment by Eliasj42** [comment]
@silvasean Hi, I just started working with the nod team.  Is there any way I could get editing permissions for this issue?

Thanks

**Comment by silvasean** [comment]
> @silvasean Hi, I just started working with the nod team. Is there any way I could get editing permissions for this issue?
> 
> Thanks

Done

**Comment by vivekkhandelwal1** [comment]
@silvasean Hi, I have started working with the Nod AI; please add me to this issue.

**Comment by silvasean** [comment]
Done!

**Comment by cathyzhyi** [comment]
@Eliasj42 could you update the ops status table with the PR numbers?

**Comment by ramiro050** [comment]
If you encounter an op that has many different variants, you can search here https://gist.github.com/ramiro050/349b61aa93a0b136bfc5c363a6a21398 to see the signature of the op being used by each of the three models

**Comment by Nirvedh** [comment]
Hi, I recently started working at Nod.ai and I am working on aten::t

**Comment by cathyzhyi** [comment]
Original op list table

| Op                                        | Status | Owner    | Model                                                 | Notes |
|-------------------------------------------|--------|----------|-------------------------------------------------------|-------|
| `aten::bmm`                               | x      |          | Bert Training                                         |       |
| `aten::gelu`                              | x      |          | Bert Training                                         |       |
| `aten::tanh`                              | x      |          | Bert Training                                         |       |
| `aten::transpose`                         | x      |          | Bert Training                                         |       |
| `aten::_softmax_backward_data`            | X      | cathyzhyi | Bert Training                                         | [PR402](https://github.com/llvm/torch-mlir/pull/402)      |
| `aten::bernoulli_`                        | -      |          | Bert Training                                         |   hold: RNG ops |
| `aten::embedding_dense_backward`          | +      | vivekkhandelwal1         | Bert Training                                         |       |
| `aten::gelu_backward`                     | x      |    gprateek93      | Bert Training                                         |  [PR418](https://github.com/llvm/torch-mlir/pull/418)     |
| `aten::index_select`                      |   x    | vivekkhandelwal1     | Bert Training                                         | https://github.com/llvm/torch-mlir/pull/466      |
| `aten::native_layer_norm_backward`        | -      |  gprateek93        | Bert Training                                         |       |
| `aten::native_layer_norm`                 | x      |  gprateek93        | Bert Training                                         |   [PR470](https://github.com/llvm/torch-mlir/pull/470)    |
| `aten::tanh_backward`                     | x    | pashu123 | Bert Training                                         | [PR408](https://github.com/llvm/torch-mlir/pull/408)      |
| `aten::permute`                           | x      |          | Bert Training, MaskRCNN Inference                     |       |
| `aten::sub`                               | x      |          | Bert Training, MaskRCNN Inference                     |       |
| `aten::unsqueeze`                         | x      |          | Bert Training, MaskRCNN Inference                     |       |
| `aten::_softmax`                          | x      |  gprateek93        | Bert Training, MaskRCNN Inference                     |   [PR431](https://github.com/llvm/torch-mlir/pull/431)   |
| `aten::select (select.int)`                            | X      |          | Bert Training, MaskRCNN Inference                     |       |
| `aten::slice(slice.Tensor)`                             | X      |          | Bert Training, MaskRCNN Inference                     |       |
| `aten::mm`                                | x      |          | Bert Training, Resnet-18 Training                     |       |
| `aten::sum`                               | x      |          | Bert Training, Resnet-18 Training                     |       |
| `aten::_log_softmax_backward_data`        | x     | pashu123         | Bert Training, Resnet-18 Training                     | [PR413](https://github.com/llvm/torch-mlir/pull/413)     |
| `aten::_log_softmax`                      | x     | pashu123 | Bert Training, Resnet-18 Training                     |  [PR397](https://github.com/llvm/torch-mlir/pull/397)     |
| `aten::addcdiv`                           | X      |     Eliasj42     | Bert Training, Resnet-18 Training                     | [PR442](https://github.com/llvm/torch-mlir/pull/442)      |
| `aten::addcmul`                           | X     |   Eliasj42     | Bert Training, Resnet-18 Training                     |  [PR442](https://github.com/llvm/torch-mlir/pull/442)     |
| `aten::nll_loss_backward`                 | +      |   pashu123       | Bert Training, Resnet-18 Training                     | [PR463]      |
| `aten::nll_loss_forward`                  | X     | pashu123 | Bert Training, Resnet-18 Training                     | [PR452]      |
| `aten::add`                               | x      |          | Bert Training, Resnet-18 Training, MaskRCNN Inference |       |
| `aten::div`                               | x      |          | Bert Training, Resnet-18 Training, MaskRCNN Inference |       |
| `aten::mul`                               | x      |          | Bert Training, Resnet-18 Training, MaskRCNN Inference |       |
| `aten::view`                              | x      |          | Bert Training, Resnet-18 Training, MaskRCNN Inference |       |
| `aten::_copy_from`                        | +      | pashu123   | Bert Training, Resnet-18 Training, MaskRCNN Inference | torchscript baseline won't run (to be investigated)|
| `aten::addmm`                             | X      |  Shukla-Gaurav       | Bert Training, Resnet-18 Training, MaskRCNN Inference |  [PR412](https://github.com/llvm/torch-mlir/pull/412)     |
| `aten::expand`                            | X      |  Shukla-Gaurav       | Bert Training, Resnet-18 Training, MaskRCNN Inference |  [PR396](https://github.com/llvm/torch-mlir/pull/396)      |
| `aten::fill_`                             | x      |  pashu123        | Bert Training, Resnet-18 Training, MaskRCNN Inference | [PR440](https://github.com/llvm/torch-mlir/pull/440)      |
| `aten::sqrt`                              | x    |          | Bert Training, Resnet-18 Training, MaskRCNN Inference |       |
| `aten::t`                                 | x      | nirvedhmeshram         | Bert Training, Resnet-18 Training, MaskRCNN Inference | [PR492](https://github.com/llvm/torch-mlir/pull/492)      |
| `aten::zero_`                             | x      | vivekkhandelwal1         | Bert Training, Resnet-18 Training, MaskRCNN Inference | https://github.com/llvm/torch-mlir/pull/464      |
| `aten::convolution_backward_overrideable` | -      |          | Resnet-18 Training                                    |       |
| `aten::max_pool2d_with_indices_backward`  | -      |          | Resnet-18 Training                                    |       |
| `aten::mean`                              | x      | vivekkhandelwal1         | Resnet-18 Training                                    | https://github.com/llvm/torch-mlir/pull/430      |
| `aten::native_batch_norm_backward`        | -      |   Shukla-Gaurav       | Resnet-18 Training                                    |       |
| `aten::native_batch_norm`                 | +      |  Shukla-Gaurav        | Resnet-18 Training                                    |       |
| `aten::random_.to`                        | -      |          | Resnet-18 Training                                    |       |
| `aten::threshold_backward`                | x      | vivekkhandelwal1         | Resnet-18 Training                                    | https://github.com/llvm/torch-mlir/pull/509      |
| `aten::relu`                              | x      |          | Resnet-18 Training, MaskRCNN Inference                |       |
| `aten::_copy_from_and_resize`             |       |          | Resnet-18 Training, MaskRCNN Inference                |       |
| `aten::convolution_overrideable`          | -      |          | Resnet-18 Training, MaskRCNN Inference                |       |
| `aten::max_pool2d_with_indices`           | +     | vivekkhandelwal1         | Resnet-18 Training, MaskRCNN Inference                | https://github.com/llvm/torch-mlir/pull/518      |
| `aten::cat`                               | x      |          | MaskRCNN Inference                                    |       |
| `aten::exp`                               | x      |          | MaskRCNN Inference                                    |       |
| `aten::sigmoid`                           | x      |          | MaskRCNN Inference                                    |       |
| `aten::clamp`                             | x      |          | MaskRCNN Inference                                    |       |
| `aten::_index_put_impl_`                  | -      |          | MaskRCNN Inference                                    |       |
| `aten::arange.start`                      | x      | vivekkhandelwal1         | MaskRCNN Inference                                    | https://github.com/llvm/torch-mlir/pull/501      |
| `aten::bitwise_and`                       | x      | vivekkhandelwal1         | MaskRCNN Inference                                    | https://github.com/llvm/torch-mlir/pull/435      |
| `aten::eq`                                | x      |      gpetters94    | MaskRCNN Inference                                    |       |
| `aten::floor`                             | x      |         | MaskRCNN Inference                                    |       |
| `aten::ge`                                | x      |    gpetters94      | MaskRCNN Inference                                    |       |
| `aten::gt`                                | +      |     gpetters94     | MaskRCNN Inference                                    |       |
| `aten::index.Tensor`                      | +      | vivekkhandelwal1         | MaskRCNN Inference                                    | https://github.com/llvm/torch-mlir/pull/495      |
| `aten::log2`                              | x      |     gpetters94     | MaskRCNN Inference                                    |       |
| `aten::lt`                                | x      |     gpetters94     | MaskRCNN Inference                                    |       |
| `aten::nonzero`                           | x     |     gpetters94     | MaskRCNN Inference                                    |       |
| `aten::rsqrt`                             | x     |     gpetters94     | MaskRCNN Inference                                    |       |
| `aten::squeeze`                           | x      |   Shukla-Gaurav       | MaskRCNN Inference                                    | [PR441](https://github.com/llvm/torch-mlir/pull/441) [PR453](https://github.com/llvm/torch-mlir/pull/453)      |
| `aten::stack`                             | -      |  pashu123        | MaskRCNN Inference                                    |       |
| `aten::topk`                              | -      |    gprateek93      | MaskRCNN Inference                                    |       |
| `aten::upsample_nearest2d`                | -      |     gprateek93     | MaskRCNN Inference                                    |       |
| `torchvision::nms`                        | -      |          | MaskRCNN Inference                                    |       |
| `torchvision::roi_align`                  | -      |          | MaskRCNN Inference                                    |       |

**Comment by powderluv** [comment]
should we trim the list to only what needs to be added will be easier to track

**Comment by cathyzhyi** [comment]
> should we trim the list to only what needs to be added will be easier to track

yes, I put the trimmed table in the original description of the issue.

**Comment by ramiro050** [comment]
Following the steps given by Chilli on discord, @pashu123 and I have been able to get a training graph for Bert into torch-mlir. Currently the lowering fails because there is no support for the following ops:

- `aten.clone`
- `aten.bernoulli_.float`
- `aten._unsafe_view`
- `aten._log_softmax`
- `aten.native_layer_norm_backward`
- `aten.embedding_dense_backward`

In case it is needed, here are the MLIR graphs showing the expected arguments for each of the functions

https://gist.github.com/pashu123/2eb1dcc5da3bccb7ca207754ec146504

https://gist.github.com/pashu123/5aff2b56a54ae00a9c177331c3eaa392

We are so close to running Bert!

**Comment by erman-gurses** [comment]
Hi everyone, I started at Nod.ai recently and I am working on aten::_index_put_impl_

**Comment by erman-gurses** [comment]
Hi everyone, How can I get edit access?

**Comment by silvasean** [comment]
> Hi everyone, How can I get edit access?

I just added you.

**Comment by gpetters94** [comment]
@ramiro050 How do you get a complete printout of missing ops? When I run MLIR through torch-mlir-opt I just get a failure on the first missing op.

**Comment by ramiro050** [comment]
> @ramiro050 How do you get a complete printout of missing ops? When I run MLIR through torch-mlir-opt I just get a failure on the first missing op.

You can usually tell which ops are missing from the MLIR, since they will have the form `torch.operator "name.of.op" ...`. However, sometimes an op that is not yet supported will not have that form. This happens when the op is added to this file (https://github.com/llvm/torch-mlir/blob/9b2613533bd78632b4438fa0b11319cd9d86f9e0/python/torch_mlir/dialects/torch/importer/jit_ir/build_tools/torch_ods_gen.py#L496) without a lowering being also added. For those ops, the easiest way to tell is just to run into error messages one op at a time.

**Comment by ramiro050** [comment]
I've generated the Bert training graph using the AOTAutograd decompositions, and the only missing ops are

- "aten.native_layer_norm_backward"
- "aten.new_empty"
- "aten.zero_"
- "aten.index_put.hacked_twin"

**Comment by silvasean** [comment]
This issue is over a year old and ResNet/BERT solidly work. MaskRCNN was moved out of scope because the model design touches on some design aspects that cross into upstream PyTorch and will be fixed as part of a longer-term plan.
